rskj-core/src/main/java/co/rsk/net/SnapshotProcessor.java

```diff
@@ -450,28 +447,34 @@ public void run() {
 
     void processSnapBlocksRequestInternal(Peer sender, SnapBlocksRequestMessage requestMessage) {
         logger.debug("Processing snap blocks request");
-        List<Block> blocks = Lists.newArrayList();
-        List<BlockDifficulty> difficulties = Lists.newArrayList();
+        LinkedList<Block> blocks = new LinkedList<>();
+        LinkedList<BlockDifficulty> difficulties = new LinkedList<>();
 
-        if (requestMessage.getBlockNumber() < 2) {
-            logger.debug("Snap blocks request from {} failed because of invalid block number {}", sender.getPeerNodeID(), requestMessage.getBlockNumber());
+        long requestBlockNumber = requestMessage.getBlockNumber();
+        if (requestBlockNumber < 2 || requestBlockNumber > blockchain.getBestBlock().getNumber()) {
+            logger.debug("Snap blocks request from {} failed because of invalid block number {}", sender.getPeerNodeID(), requestBlockNumber);
             return;
         }
 
-        long startingBlockNumber = Math.max(1, requestMessage.getBlockNumber() - BLOCK_CHUNK_SIZE);
-        for (long i = startingBlockNumber; i < requestMessage.getBlockNumber(); i++) {
-            Block block = blockchain.getBlockByNumber(i);
-            if (block == null) {
-                break;
-            }
-            blocks.add(block);
-            difficulties.add(blockStore.getTotalDifficultyForHash(block.getHash().getBytes()));
-        }
+        long fromBlock = Math.max(1, requestMessage.getBlockNumber() - BLOCK_CHUNK_SIZE);
+        retrieveBlocksAndDifficultiesBackwards(fromBlock,requestBlockNumber-1, blocks, difficulties);
         logger.debug("Sending snap blocks response. From block {} to block {} - chunksize {}", blocks.get(0).getNumber(), blocks.get(blocks.size() - 1).getNumber(), BLOCK_CHUNK_SIZE);
         SnapBlocksResponseMessage responseMessage = new SnapBlocksResponseMessage(requestMessage.getId(), blocks, difficulties);
         sender.sendMessage(responseMessage);
     }
 
+    private void retrieveBlocksAndDifficultiesBackwards(long fromBlock, long toBlock, LinkedList<Block> blocks, LinkedList<BlockDifficulty> difficulties) {
+        Block currentBlock = blockchain.getBlockByNumber(toBlock);
+        while (currentBlock != null && currentBlock.getNumber() >= fromBlock) {
+            blocks.addFirst(currentBlock);
+            difficulties.addFirst(blockStore.getTotalDifficultyForBlock(currentBlock));
+            currentBlock = blockStore.getBlockByHash(currentBlock.getParentHash().getBytes());
+        }
+        if (currentBlock == null) {
+            logger.warn("No block found for block number {}", toBlock);
```

Comment:
in some cases `toBlock` may not refer to a missing block, eg. if `blockStore.getBlockByHash(currentBlock.getParentHash().getBytes())` returns null

Reply:
my bad, I forgot that.